tree/ntuple/v7/src/RNTupleProcessor.cxx

```diff
@@ -223,34 +232,45 @@ ROOT::Experimental::NTupleSize_t ROOT::Experimental::RNTupleChainProcessor::Conn
    fPageSource = Internal::RPageSource::Create(ntuple.fNTupleName, ntuple.fStorage);
    fPageSource->Attach();
 
+   fInnerNEntries[ntupleNumber] = fPageSource->GetNEntries();
+
    // Now that the new page source has been created and attached, we can create and connect the concrete fields again.
    for (auto &[_, fieldContext] : fFieldContexts) {
       ConnectField(fieldContext, *fPageSource, *fEntry);
    }
 
+   fCurrentNTupleNumber = ntupleNumber;
+
    return fPageSource->GetNEntries();
 }
 
-ROOT::Experimental::NTupleSize_t ROOT::Experimental::RNTupleChainProcessor::Advance()
+ROOT::Experimental::NTupleSize_t ROOT::Experimental::RNTupleChainProcessor::LoadEntry(NTupleSize_t entryNumber)
 {
-   if (fLocalEntryNumber == kInvalidNTupleIndex)
-      return kInvalidNTupleIndex;
+   NTupleSize_t localEntryNumber = entryNumber;
+   size_t currentNTuple = 0;
 
-   if (fLocalEntryNumber >= fPageSource->GetNEntries()) {
-      do {
-         if (++fCurrentNTupleNumber >= fNTuples.size()) {
-            return kInvalidNTupleIndex;
-         }
-         // Skip over empty ntuples we might encounter.
-      } while (ConnectNTuple(fNTuples.at(fCurrentNTupleNumber)) == 0);
+   while (localEntryNumber >= fInnerNEntries[currentNTuple]) {
+      localEntryNumber -= fInnerNEntries[currentNTuple];
+
+      if (++currentNTuple >= fNTuples.size())
+         return kInvalidNTupleIndex;
 
-      fLocalEntryNumber = 0;
+      if (fInnerNEntries[currentNTuple] == kInvalidNTupleIndex) {
+         auto tmpPageSource =
+            Internal::RPageSource::Create(fNTuples[currentNTuple].fNTupleName, fNTuples[currentNTuple].fStorage);
+         tmpPageSource->Attach();
+         fInnerNEntries[currentNTuple] = tmpPageSource->GetNEntries();
```

Comment:
For future optimizations:
1. If we actually find the entry in this page source, we should not destroy it. On the other hand, we don't need the full `ConnectNTuple` call for intermediate ntuples that we just skip over.
2. If you store the cumulative number of entries, you can
a. check if `entryNumber` is smaller than the highest cached value in `fInnerNEntries`, only if not you have to linearly go through the ntuples.
b. If the `entryNumber` is contained, you can do binary search in the cached `fInnerNEntries`

(doesn't need to be in this PR)